File: BasicPlaylist/src/main/java/com/bitmovin/player/samples/playlist/basic/MainActivity.java
package com.bitmovin.player.samples.playlist.basic;

import android.os.Bundle;
import android.view.View;
import android.view.Window;
import android.widget.Toast;

import androidx.activity.EdgeToEdge;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.core.view.WindowInsetsControllerCompat;

import com.bitmovin.player.PlayerView;
import com.bitmovin.player.api.Player;
import com.bitmovin.player.api.event.EventListener;
import com.bitmovin.player.api.event.PlayerEvent;
import com.bitmovin.player.api.playlist.PlaylistApi;
import com.bitmovin.player.api.playlist.PlaylistConfig;
import com.bitmovin.player.api.playlist.PlaylistOptions;
import com.bitmovin.player.api.source.Source;
import com.bitmovin.player.api.source.SourceBuilder;
import com.bitmovin.player.api.source.SourceConfig;
import com.bitmovin.player.api.source.SourceType;

import java.util.Arrays;
import java.util.List;


public class MainActivity extends AppCompatActivity {
    private PlayerView playerView;
    private Player player;

    private static final String SintelHls = "https://cdn.bitmovin.com/content/assets/sintel/hls/playlist.m3u8";
    private static final String ArtOfMotionProgressive = "https://cdn.bitmovin.com/content/assets/MI201109210084/MI201109210084_mpeg-4_hd_high_1080p25_10mbits.mp4";
    private static final String SintelDash = "https://cdn.bitmovin.com/content/assets/sintel/sintel.mpd";
    private static final String KronehitLiveHls = "https://bitcdn-kronehit.bitmovin.com/v2/hls/playlist.m3u8";

    private final EventListener<PlayerEvent.PlaylistTransition> onReadyListener = event -> {
        String text = "Transitioned from " + event.getFrom().getConfig().getTitle() + " to " + event.getTo().getConfig().getTitle();
        Toast.makeText(MainActivity.this, text, Toast.LENGTH_SHORT).show();
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        EdgeToEdge.enable(this);
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.root), (view, insets) -> {
            Insets bars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            view.setPadding(bars.left, bars.top, bars.right, bars.bottom);
            return WindowInsetsCompat.CONSUMED;
        });
        Window window = getWindow();
        WindowInsetsControllerCompat insetsController = WindowCompat.getInsetsController(window,
                window.getDecorView());
        insetsController.setAppearanceLightStatusBars(true);
        insetsController.setAppearanceLightNavigationBars(true);

        this.playerView = this.findViewById(R.id.playerView);
        this.player = this.playerView.getPlayer();

        SourceConfig sourceConfig1 = new SourceConfig(SintelHls, SourceType.Hls);
        sourceConfig1.setTitle("(1/4) Sintel HLS");
        SourceConfig sourceConfig2 = new SourceConfig(ArtOfMotionProgressive, SourceType.Progressive);
        sourceConfig2.setTitle("(2/4) Art of Motion Progressive");
        SourceConfig sourceConfig3 = new SourceConfig(SintelDash, SourceType.Dash);
        sourceConfig3.setTitle("(3/4) Sintel DASH");
        SourceConfig sourceConfig4 = new SourceConfig(KronehitLiveHls, SourceType.Hls);
        sourceConfig4.setTitle("(4/4) Kronehit Live HLS");

        List<Source> sources = Arrays.asList(
                new SourceBuilder(sourceConfig1).build(),
                new SourceBuilder(sourceConfig2).build(),
                new SourceBuilder(sourceConfig3).build(),
                new SourceBuilder(sourceConfig4).build()
        );

        this.player.on(PlayerEvent.PlaylistTransition.class, this.onReadyListener);
        this.player.load(new PlaylistConfig(sources, new PlaylistOptions()));
    }


    @Override
    protected void onStart() {
        this.playerView.onStart();
        super.onStart();
    }

    @Override
    protected void onResume() {
        super.onResume();
        this.playerView.onResume();
    }

    @Override
    protected void onPause() {
        this.playerView.onPause();
        super.onPause();
    }

    @Override
    protected void onStop() {
        this.playerView.onStop();
        super.onStop();
    }

    @Override
    protected void onDestroy() {
        this.playerView.onDestroy();
        super.onDestroy();
    }

    /**
     * @return the index of the current active source or -1 if none is active
     */
    private int getActiveSourceIndex(PlaylistApi playlist) {
        List<Source> sources = playlist.getSources();
        for (int index = 0; index < sources.size(); index++) {
            if (sources.get(index).isActive()) {
                return index;
            }
        }
        return -1;
    }

    public void next(View view) {
        playlistIncrement(1);
    }

    public void previous(View view) {
        playlistIncrement(-1);
    }

    private void playlistIncrement(int increment) {
        final PlaylistApi playlist = this.player.getPlaylist();
        final int activeSourceIndex = getActiveSourceIndex(playlist);
        if (activeSourceIndex == -1) {
            return;
        }
        final int newSourceIndex = activeSourceIndex + increment;
        if (newSourceIndex < 0 || newSourceIndex >= playlist.getSources().size()) {
            return;
        }

        playlist.seek(playlist.getSources().get(newSourceIndex), 0.0);
    }
}

File: BasicPlaylistKotlin/src/main/java/com/bitmovin/player/samples/playlist/basic/MainActivity.kt
package com.bitmovin.player.samples.playlist.basic

import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.updatePadding
import com.bitmovin.analytics.api.AnalyticsConfig
import com.bitmovin.player.api.Player
import com.bitmovin.player.api.analytics.AnalyticsPlayerConfig
import com.bitmovin.player.api.event.PlayerEvent
import com.bitmovin.player.api.event.SourceEvent
import com.bitmovin.player.api.event.next
import com.bitmovin.player.api.event.on
import com.bitmovin.player.api.playlist.PlaylistApi
import com.bitmovin.player.api.playlist.PlaylistConfig
import com.bitmovin.player.api.playlist.PlaylistOptions
import com.bitmovin.player.api.playlist.ReplayMode
import com.bitmovin.player.api.source.Source
import com.bitmovin.player.api.source.SourceConfig
import com.bitmovin.player.api.source.SourceType
import com.bitmovin.player.samples.playlist.basic.databinding.ActivityMainBinding

private const val SintelHls = "https://cdn.bitmovin.com/content/assets/sintel/hls/playlist.m3u8"
private const val ArtOfMotionProgressive = "https://cdn.bitmovin.com/content/assets/MI201109210084/MI201109210084_mpeg-4_hd_high_1080p25_10mbits.mp4"
private const val SintelDash = "https://cdn.bitmovin.com/content/assets/sintel/sintel.mpd"
private const val KronehitLiveHls = "https://bitcdn-kronehit.bitmovin.com/v2/hls/playlist.m3u8"

class MainActivity : AppCompatActivity() {
    private lateinit var player: Player
    private lateinit var binding: ActivityMainBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        enableEdgeToEdge()
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        ViewCompat.setOnApplyWindowInsetsListener(binding.root) { view, windowInsets ->
            val insets = windowInsets.getInsets(WindowInsetsCompat.Type.systemBars())
            view.updatePadding(insets.left, insets.top, insets.right, insets.bottom)
            WindowInsetsCompat.CONSUMED
        }
        WindowCompat.getInsetsController(window, window.decorView).apply {
            isAppearanceLightStatusBars = true
            isAppearanceLightNavigationBars = true
        }

        val analyticsKey = "{ANALYTICS_LICENSE_KEY}"
        player = Player(
            context = this,
            analyticsConfig = AnalyticsPlayerConfig.Enabled(AnalyticsConfig(analyticsKey)),
        ).also {
            binding.playerView.player = it
        }

        initializePlayer()
    }

    override fun onStart() {
        super.onStart()
        binding.playerView.onStart()
    }

    override fun onResume() {
        super.onResume()
        binding.playerView.onResume()
    }

    override fun onPause() {
        binding.playerView.onPause()
        super.onPause()
    }

    override fun onStop() {
        binding.playerView.onStop()
        super.onStop()
    }

    override fun onDestroy() {
        binding.playerView.onDestroy()
        player.off(::onPlaylistTransition)
        super.onDestroy()
    }

    private fun initializePlayer() {
        val sources = listOf(
                Source(
                        SourceConfig(
                                url = SintelHls,
                                type = SourceType.Hls,
                                title = "(1/4) Sintel HLS",
                        )
                ),
                Source(
                        SourceConfig(
                                url = ArtOfMotionProgressive,
                                type = SourceType.Progressive,
                                title = "(2/4) Art of Motion Progressive",
                        )
                ),
                Source(
                        SourceConfig(
                                url = SintelDash,
                                type = SourceType.Dash,
                                title = "(3/4) Sintel DASH",
                        )
                ),
                Source(
                        SourceConfig(
                                url = KronehitLiveHls,
                                type = SourceType.Hls,
                                title = "(4/4) Kronehit Live HLS",
                        )
                )
        )

        sources.forEach { source ->
            source.next<SourceEvent.Loaded> {
                println("Loaded ${it.source.config.title} with duration ${it.source.duration}")
            }
        }

        val playlistConfig = PlaylistConfig(
                sources,
                PlaylistOptions(
                        preloadAllSources = true,
                        replayMode = ReplayMode.Playlist
                )
        )

        player.on(::onPlaylistTransition)

        player.load(playlistConfig)
    }

    private fun onPlaylistTransition(event: PlayerEvent.PlaylistTransition) {
        binding.playerView.post {
            val text = "Transitioned from ${event.from.config.title} to ${event.to.config.title}."
            Toast.makeText(this, text, Toast.LENGTH_SHORT).show()
        }
    }

    fun previous(view: View) = player.playlist.previous()

    fun next(view: View) = player.playlist.next()
}


fun PlaylistApi.next(time: Double = 0.0) {
    (sources.indexOf(activeSource) + 1)
            .takeIf { it < sources.count() }
            ?.let { seek(sources[it], time) }
}

fun PlaylistApi.previous(time: Double = 0.0) {
    (sources.indexOf(activeSource) - 1)
            .takeIf { it > -1 }
            ?.let { seek(sources[it], time) }
}

private val PlaylistApi.activeSource: Source? get() = sources.firstOrNull { it.isActive }
